libraries/networking/src/PacketReceiver.cpp

```diff
@@ -0,0 +1,229 @@
+                                               senderSockAddr.getAddressPointer(), senderSockAddr.getPortPointer());
+
+        // setup an NLPacket from the data we just read
+        auto packet = NLPacket::fromReceivedPacket(std::move(buffer), packetSizeWithHeader, senderSockAddr);
+        
+        _inPacketCount++;
+        _inByteCount += packetSizeWithHeader;
+
+        if (packetVersionMatch(*packet)) {
+            
+            SharedNodePointer matchingNode;
+            if (nodeList->packetSourceAndHashMatch(*packet, matchingNode)) {
+                 
+                if (matchingNode) {
+                    // No matter if this packet is handled or not, we update the timestamp for the last time we heard
+                    // from this sending node
+                    matchingNode->setLastHeardMicrostamp(usecTimestampNow());
+                }
+
+                _packetListenerLock.lock();
+                
+                auto it = _packetListenerMap.find(packet->getType());
+
+                if (it != _packetListenerMap.end()) {
+
+                    auto listener = it.value();
+                   
+                    if (!listener.first.isNull()) {
+                        
+                        if (matchingNode) {
+                            emit dataReceived(matchingNode->getType(), packet->getSizeWithHeader());
+                        } else {
+                            emit dataReceived(NodeType::Unassigned, packet->getSizeWithHeader());
+                        }
+                        
+                        bool success = false;
+
+                        if (matchingNode) {
+                            success = listener.second.invoke(listener.first,
+                                Q_ARG(QSharedPointer<NLPacket>, QSharedPointer<NLPacket>(packet.release())));
+                        } else {
+                            success = listener.second.invoke(listener.first,
+                                Q_ARG(QSharedPointer<NLPacket>, QSharedPointer<NLPacket>(packet.release())),
+                                Q_ARG(SharedNodePointer, matchingNode));
+                        }
+                                            
+                        if (!success) {
+                            qDebug() << "Error delivering packet " << nameForPacketType(packet->getType()) << " to listener: "
+                                << listener.first->objectName() << "::" << listener.second.name();
+                        }
+
+                    } else {
+                        // we have a dead listener - remove this mapping from the _packetListenerMap
+                        qDebug() << "Listener for packet type" << nameForPacketType(packet->getType())
+                            << "has been destroyed - removing mapping.";
+                        _packetListenerMap.erase(it);
+                    }
+
+                     _packetListenerLock.unlock();
```

Comment:
extra space at beginning, although I think we should just move this unlock below the `if/else` (and also remove it from the `else` block)